packages/node/src/plugins/autostaker/AutostakerPlugin.ts

```diff
@@ -0,0 +1,87 @@
+import { _operatorContractUtils, StreamrClient } from '@streamr/sdk'
+import { collect, Logger, scheduleAtInterval, WeiAmount } from '@streamr/utils'
+import { Schema } from 'ajv'
+import { formatEther, parseEther, Wallet } from 'ethers'
+import sample from 'lodash/sample'
+import { Plugin } from '../../Plugin'
+import PLUGIN_CONFIG_SCHEMA from './config.schema.json'
+
+export interface AutostakerPluginConfig {
+    operatorContractAddress: string
+    // TODO is it possible implement this without exposing the private key here?
+    // e.g. by configuring so that operator nodes can stake behalf of the operator?
+    operatorOwnerPrivateKey: string
+    runIntervalInMs: number
+}
+
+const STAKE_AMOUNT: WeiAmount = parseEther('10000')
+
+const logger = new Logger(module)
+
+export class AutostakerPlugin extends Plugin<AutostakerPluginConfig> {
+
+    private abortController: AbortController = new AbortController()
+
+    async start(streamrClient: StreamrClient): Promise<void> {
+        logger.info('Start autostaker plugin')
+        scheduleAtInterval(async () => {
+            try {
+                await this.runActions(streamrClient)
+            } catch (err) {
+                logger.warn('Error while running autostaker actions', { err })
+            }
+        }, this.pluginConfig.runIntervalInMs, false, this.abortController.signal)
+    }
+
+    private async runActions(streamrClient: StreamrClient): Promise<void> {
+        logger.info('Run autostaker actions')
+        const provider = (await streamrClient.getSigner()).provider
+        const operatorContract = _operatorContractUtils.getOperatorContract(this.pluginConfig.operatorContractAddress)
+            .connect(new Wallet(this.pluginConfig.operatorOwnerPrivateKey, provider))
+        const stakedAmount = await operatorContract.totalStakedIntoSponsorshipsWei()
+        const availableBalance = (await operatorContract.valueWithoutEarnings()) - stakedAmount
+        logger.info(`Available balance: ${formatEther(availableBalance)} (staked=${formatEther(stakedAmount)})`)
+        // TODO is there a better way to get the client? Maybe we should add StreamrClient#getTheGraphClient()
+        // TODO what are good where consitions for the sponsorships query
+        // @ts-expect-error private
+        const queryResult = streamrClient.theGraphClient.queryEntities((lastId: string, pageSize: number) => {
+            return {
+                query: `
```

Comment:
> How we should query the possible sponsorships, which we could stake to? Now we use projectedInsolvency_gt and spotAPY_gt fllters, but maybe these are not optimall.

`remainingWei_gt: "0"` is same as the projectedInsolvency, and clearer IMO. It's a good way to select only "active" sponsorships.

There might not need to be hard cutoffs for other this, just order by `totalPayoutWeiPerSec` desc should be fine, take some reasonable maximum number, e.g. 100.

Reply:
On the other hand, in tests, remainingWei isn't updated instantly; in a test, we can do a "zero sponsoring" to update the remainingWei (after it should be expired).

Reply:
The current `projectedInsolvency_gt: ${Math.floor(Date.now() / 1000)}` is in other words totally fine, and perhaps it's easier to write the test as well. I'll mark this conversation resolved.